apps/fft/fft_aot_test.cpp

```diff
@@ -0,0 +1,288 @@
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+
+#include "generated/fft_forward_r2c.h"
+#include "generated/fft_inverse_c2r.h"
+#include "generated/fft_forward_c2c.h"
+#include "generated/fft_inverse_c2c.h"
+
+namespace {
+const float kPi = 3.14159265358979310000f;
+
+const size_t kSize = 16;
+}
+
+// Make a buffer_t for real input to the FFT.
+buffer_t real_buffer(float *storage, int32_t y_size = kSize) {
+    buffer_t buf = {0};
+
+    buf.host = (uint8_t *)storage;
+    buf.extent[0] = kSize;
+    buf.stride[0] = 1;
+    buf.extent[1] = y_size;
+    buf.stride[1] = kSize;
+    buf.elem_size = sizeof(float);
+
+    return buf;
+}
+
+// Make a buffer_t for complex input to the FFT.
+buffer_t complex_buffer(float *storage, int32_t y_size = kSize) {
+    buffer_t buf = {0};
+
+    buf.host = (uint8_t *)storage;
+    buf.extent[0] = 2;
```

Comment:
The main reason I avoided Tuples is because it changes the number of arguments to the filter depending on whether an input or output is real or complex. This is fine for output, but it runs up against an issue with Generators on the input side in that Generator does not allow an argument to sometimes exist and sometimes not depending on GeneratorParams. So I have to either require an unused input argument for the real case or I have to register multiple generators and make subclasses. The latter is doable, but I'd rather add it later if need be. Steven and I have discussed adding Generator support for arguments that are sometimes there and sometimes not, but all of that plays into the design for Generator fusion and frankly for right now, if you need to do that stuff, the C++ fft.h interface is likely a better way to go.

Beyond that, most FFT packages provide the packed representation as their main thing so we definitely need to provide this interface. They may support separating the components as well, but I believe usually having the complex components next to each other is preferred. Note that it is kind of like packed vs. planar for pixel channel components, but not quite as usually the complex components are completely separate when they are not together. (Since there are only two of them, one can model this with a plane stride, but it is weird.)

The idea behind using the inner most component is both to allow processing multiple FFTs on all the trailing dimensions and the idea that complex is a fundamental type. The latter is basically true in C99 and C++ actually. E.g. One option is taking ComplexFunc and just building it into the Halide type system and expression language mapping to the corresponding C/C++ types. (And while I don't have a strong opinion on this, I don't think it is a bad idea.)